offscreen/scripts/offscreen.ts

```diff
@@ -9,41 +10,63 @@ import initLedger from './ledger';
 import initTrezor from './trezor';
 import initLattice from './lattice';
 
-initLedger();
-initTrezor();
-initLattice();
-
 /**
  * Initialize a post message stream with the parent window that is initialized
  * in the metamask-controller (background/serivce worker) process. This will be
  * utilized by snaps for communication with snaps running in the offscreen
  * document.
  */
-const parentStream = new BrowserRuntimePostMessageStream({
-  name: 'child',
-  target: 'parent',
-});
-
-ProxySnapExecutor.initialize(parentStream, './snaps/index.html');
-
-if (process.env.IN_TEST) {
-  chrome.runtime.onMessage.addListener((message) => {
-    if (
-      message &&
-      typeof message === 'object' &&
-      message.event === OffscreenCommunicationEvents.metamaskBackgroundReady &&
-      message.target === OffscreenCommunicationTarget.extension
-    ) {
-      window.document?.documentElement?.classList?.add('controller-loaded');
-    }
+function initializePostMessageStream() {
+  const parentStream = new BrowserRuntimePostMessageStream({
+    name: 'child',
+    target: 'parent',
   });
+
+  ProxySnapExecutor.initialize(parentStream, './snaps/index.html');
 }
 
-chrome.runtime.sendMessage({
-  target: OffscreenCommunicationTarget.extensionMain,
-  isBooted: true,
+/**
+ * Initialize the ledger, trezor, and lattice keyring connections, and the
+ * post message stream for the Snaps environment.
+ */
+async function init(): Promise<void> {
+  initializePostMessageStream();
+  initTrezor();
+  initLattice();
```

Comment:
Just out of curiosity, would I be correct in assuming that the post message stream needs to be synchronously the first task to complete, and then the order doesn't matter between trezor, lattice, and ledger?

Would there be any potential performance, feature, or security gain from initializing the hardware wallets concurrently? Feel free to disregard if this doesn't make any sense, since I don't have much of an understanding of this domain.

Reply:
They are effectively executed concurrently in this case, because we're not `await`-ing anything here. Good question though, in case we did want to `await` these other steps as well for similar reasons as for Ledger.

Reply:
Oh that's super cool.😎 I believe these 3 lines would be processed by the call stack instead of the task queue so they would at least always be executed sequentially and in order. I wonder if we could get true parallelism with web workers for this or other similar collections of tasks where speed gains matter. Interesting food for thought!

Reply:
I believe they all either spawn an iframe in the offscreen page (i.e. Snaps, Ledger) or setup page event listeners (i.e. Trezor, Lattice): they are all executed sequentially, though we don't wait for the Snap iframe to be fully loaded to move on and init the rest.

The order is not strict, but I thought it made sense to load Ledger as last since it is the only one we explicitly wait for, and this gives a chance to the Swap iframe to load in the meantime - while if we, for example, were to load Ledger as first, the rest would wait for it to either resolve or timeout.

Reply:
> Good question though, in case we did want to await these other steps as well for similar reasons as for Ledger

Since Trezor and Lattice only require an event listener to be set on `chrome.runtime`, they are effectively already a sync step that we can't really parallelize.

The only component we are initializing out of band is `ProxySnapExecutor`

Reply:
It would be nice to explore the possibility of only setting an event listener for Ledger as well, and spawning the iframe later in time only in case a Ledger Keyring is added to the wallet when the iframe is really needed - because we currently add a Ledger iframe and a LedgerKeyring to the wallet even if the user never interacted with a ledger device